consensus/wrapper/consensus.go

```diff
@@ -151,34 +151,77 @@ func (ue *UpgradableEngine) VerifyHeaders(chain consensus.ChainHeaderReader, hea
 	return abort, results
 }
 
+func (ue *UpgradableEngine) CalcTimestamp(parent *types.Header) uint64 {
+	panic("Called CalcTimestamp on UpgradableEngine, not implemented")
+}
+
 // Prepare prepares a block header for sealing.
-func (ue *UpgradableEngine) Prepare(chain consensus.ChainHeaderReader, header *types.Header) error {
-	return ue.chooseEngine(header).Prepare(chain, header)
+func (ue *UpgradableEngine) Prepare(chain consensus.ChainHeaderReader, header *types.Header, timeOverride *uint64) error {
+	// Override provided => select engine based on override timestamp.
+	if timeOverride != nil {
+		return ue.chooseEngine(*timeOverride).Prepare(chain, header, timeOverride)
+	}
+
+	parent := chain.GetHeader(header.ParentHash, header.Number.Uint64()-1)
+	if parent == nil {
+		return consensus.ErrUnknownAncestor
+	}
+
+	// Check if parent is pre- or post-EuclidV2.
+	if ue.chooseEngine(parent.Time) == ue.clique {
+		// This is either a normal Clique block, or the EuclidV2 transition block.
+
+		time := ue.clique.CalcTimestamp(parent)
+
+		if ue.chooseEngine(time) == ue.clique {
+			// We're still in Clique mode.
+			// Note: We must provide timestamp override to avoid the edge case
+			// where we slip over into EuclidV2 in this next call.
+			return ue.clique.Prepare(chain, header, &time)
+		} else {
+			// This is the EuclidV2 transition block.
+			return ue.system.Prepare(chain, header, &time)
+		}
+	} else {
+		// We are already post EuclidV2, in SystemContract mode.
+
+		time := ue.system.CalcTimestamp(parent)
+
+		if ue.chooseEngine(time) == ue.clique {
+			// Somehow we slipped back to Clique, override with a known post-EuclidV2 timestamp.
+			// Note: This should not happen in practice.
+			log.Error("Parent is post-EuclidV2 but SystemContract set pre-EuclidV2 timestamp, overriding", "blockNumber", header.Number, "parentTime", parent.Time, "systemContractTime", time)
+			return ue.system.Prepare(chain, header, &parent.Time)
```

Comment:
Same time as the parent? Not even slightly greater?

Reply:
Same time, as implemented by @jonastheis. I don't see any issue with this, both engines allow this. And this is an edge case that probably won't be triggered. but just in case.